src/plot.cpp

```diff
@@ -1,10 +1,1034 @@
+    i.resize(height);
+    for (auto & j : i) {
+      j.resize(3);
+    }
+  }
+
+  int in_i, out_i;
+  double xyz[3];
+  switch(pl->basis) {
+  case PLOT_BASIS::XY :
+    in_i = 0;
+    out_i = 1;
+    xyz[0] = pl->origin[0] - pl->width[0] / TWO;
+    xyz[1] = pl->origin[1] + pl->width[1] / TWO;
+    xyz[2] = pl->origin[2];
+    break;
+  case PLOT_BASIS::XZ :
+    in_i = 0;
+    out_i = 2;
+    xyz[0] = pl->origin[0] - pl->width[0] / TWO;
+    xyz[1] = pl->origin[1];
+    xyz[2] = pl->origin[2] + pl->width[1] / TWO;
+    break;
+  case PLOT_BASIS::YZ :
+    in_i = 1;
+    out_i = 2;
+    xyz[0] = pl->origin[0];
+    xyz[1] = pl->origin[1] - pl->width[0] / TWO;
+    xyz[2] = pl->origin[2] + pl->width[1] / TWO;
+    break;
+  }
+
+  double dir[3] = {HALF, HALF, HALF};
+  Particle *p = new Particle();
+  p->initialize();
+  std::copy(xyz, xyz+3, p->coord[0].xyz);
+  std::copy(dir, dir+3, p->coord[0].uvw);
+  p->coord[0].universe = openmc_root_universe;
+
+  // local variables
+  int rgb[3];
+  int id;
+  for (int y = 0; y < height; y++) {
+    p->coord[0].xyz[out_i] = xyz[out_i] - out_pixel*(y);
+    for (int x = 0; x < width; x++) {
+      p->coord[0].xyz[in_i] = xyz[in_i] + in_pixel*(x);
+      position_rgb(p, pl, rgb, id);
+      data[x][y][RED] = rgb[RED];
+      data[x][y][GREEN] = rgb[GREEN];
+      data[x][y][BLUE] = rgb[BLUE];
+    }
+  }
+
+  // draw mesh lines if present
+  if (pl->index_meshlines_mesh >= 0) { draw_mesh_lines(pl, data); }
+
+  // write ppm data to file
+  output_ppm(pl, data);
+}
+
+void
+ObjectPlot::set_id()
+{
+    // Copy data into plots
+  if (check_for_node(_plot_node, "id")) {
+    id = std::stoi(get_node_value(_plot_node, "id"));
+  } else {
+    fatal_error("Must specify plot id in plots XML file.");
+  }
+
+  // Check to make sure 'id' hasn't been used
+  if (plot_dict.find(id) != plot_dict.end()) {
+    std::stringstream err_msg;
+    err_msg << "Two or more plots use the same unique ID: ";
+    err_msg << id;
+    fatal_error(err_msg.str());
+  }
+}
+
+void
+ObjectPlot::set_type()
+{
+  // Copy plot type
+  // Default is slice
+  std::string type_str = "slice";
+  type = PLOT_TYPE::SLICE;
+  // check type specified on plot node
+  if (check_for_node(_plot_node, "type")) {
+    type_str = get_node_value(_plot_node, "type");
+    to_lower(type_str);
+    // set type using node value
+    if (type_str == "slice") {
+      type = PLOT_TYPE::SLICE;
+      return;
+    }
+    else if (type_str == "voxel") {
+      type = PLOT_TYPE::VOXEL;
+      return;
+    }
+    // if we're here, something is wrong
+    std::stringstream err_msg;
+    err_msg << "Unsupported plot type '" << type_str
+            << "' in plot " << id;
+    fatal_error(err_msg.str());
+  }
+}
+
+void
+ObjectPlot::set_output_path()
+{
+  // Set output file path
+  std::stringstream filename;
+  filename << "plot_" << id;
+
+  if (check_for_node(_plot_node, "filename")) {
+    filename << get_node_value(_plot_node, "filename");
+  } else {
```

Comment:
Extensions get added below before adding them to `path_plot` I think... am I missing something?

Reply:
I guess what I'm trying to say is that the else line is not needed, and the if statement before will append `filename` to `plot_ << id` 

For example, if you change line 26 of `plots.xml` in `openmc/tests/regression_tests/plot` to `<plot id="4" type="voxel" filename="temp">`, the resulting output file will be `plot_4temp` instead of `temp.h5`, which is what the Fortran version currently would output.

Reply:
Ah! Totally. Missed that else for some reason. Thank you. Fixing...